File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/send/core/UserOpSpine.kt
package app.getvela.wallet.feature.send.core

import app.getvela.wallet.core.diagnostics.VelaLog
import app.getvela.wallet.feature.onboarding.core.Assertion
import app.getvela.wallet.feature.onboarding.core.FailureKind
import app.getvela.wallet.feature.onboarding.core.KeyMethod
import app.getvela.wallet.feature.onboarding.core.PasskeyFailure
import org.json.JSONObject
import uniffi.vela_core_uniffi.RelayRejection
import uniffi.vela_core_uniffi.WalletKeyRecord
import uniffi.vela_core_uniffi.UserOpCall
import uniffi.vela_core_uniffi.UserOpDraft
import uniffi.vela_core_uniffi.userOpCallsToMeasure
import uniffi.vela_core_uniffi.userOpRaiseCallGas
import uniffi.vela_core_uniffi.UserOpFeeMode
import uniffi.vela_core_uniffi.WebAuthnAssertion
import uniffi.vela_core_uniffi.eip1271Signature
import uniffi.vela_core_uniffi.safeMessageHash
import uniffi.vela_core_uniffi.classifyRelayRejection
import uniffi.vela_core_uniffi.isChainWithoutNativeCoin
import uniffi.vela_core_uniffi.parseExistingUserOpHash
import uniffi.vela_core_uniffi.quotedFeeUsable
import uniffi.vela_core_uniffi.relayErrorMessage
import uniffi.vela_core_uniffi.userOpApplyEstimate
import uniffi.vela_core_uniffi.userOpDraft
import uniffi.vela_core_uniffi.userOpFloors
import uniffi.vela_core_uniffi.userOpHasContractCall
import uniffi.vela_core_uniffi.userOpRelayJson
import uniffi.vela_core_uniffi.userOpSafeOpHash
import uniffi.vela_core_uniffi.userOpSign
import uniffi.vela_core_uniffi.userOpWithCalls

/**
 * The submit spine — ONE implementation for a person's own transfer and a
 * dApp's transaction (spec 044 T028, research D7; the desktop's
 * `executor::user_op::submit`).
 *
 * The order is `sendUserOpInBand`'s: isDeployed → nonce → floors → the
 * placeholder fee leg → estimate → apply → the settled fee leg → the SafeOp
 * hash is the challenge → the assertion → the envelope → the relay. The
 * assembly is the core's (`user_op_*` through uniffi); the transports are
 * [RelayClient]'s; the one seam the core cannot have is [UserOpSigner].
 * Nothing here prices, validates or classifies beyond what the core exports.
 *
 * Extracted from `SendExecutor` so the sheet that shows what will happen
 * and the code that makes it happen cannot become two opinions.
 */
class UserOpSpine(
    private val relay: RelayClient,
    private val accounts: SendExecutor.AccountPort,
    private val signer: () -> UserOpSigner,
    /**
     * `eth_estimateGas({from, to, value, data})` on one chain — the gas hex, or
     * `null` when nobody answered. The inner calls' own measurement (spec 062):
     * the relay estimates `callGasLimit` against the SENDER, and an undeployed
     * Safe has no code there, so its figure is 21k plus calldata whatever the
     * call does. Measured from the Safe's address instead, a codeless account
     * estimates like any other. The rule is the core's (`userOpRaiseCallGas`).
     */
    private val measureCall: suspend (chainId: Int, from: String, to: String, valueHex: String, data: String) -> String? =
        { _, _, _, _, _ -> null },
    /**
     * The person's "Sign with" choice for the request in hand — `auto` unless a
     * signing sheet says otherwise. WHICH key that pins, and how it is reached,
     * is the core's (`signRoute`); `auto` is the stored route, untouched.
     */
    private val signMethod: () -> String = { "auto" },
    private val route: (keysJson: String, method: String) -> String? = { keys, method -> uniffi.vela_core_uniffi.signRoute(keys, method) },
) {
    /** The credential a ceremony is pinned to, its transports and method. */
    private suspend fun routeFor(account: String, first: WalletKeyRecord): Triple<String, String, KeyMethod> {
        val chosen = signMethod()
        if (chosen != "auto") {
            runCatching { route(accounts.keyRoutesJson(account), chosen)?.let(::JSONObject) }.getOrNull()?.let { picked ->
                val method = KeyMethod.entries.firstOrNull { it.wire == picked.optString("method") }
                if (method != null && picked.optString("credential_id").isNotEmpty()) {
                    return Triple(picked.optString("credential_id"), picked.optString("transports"), method)
                }
            }
        }
        val (transports, method) = accounts.routingOf(account)
        return Triple(first.credentialId, transports, method)
    }

    /** The displayed fee, signed verbatim. */
    data class Quoted(val amount: String, val recipient: String)

    sealed class Failure {
        data object PasskeyCancelled : Failure()
        data object RelayerUnavailable : Failure()
        data object BundlerUnderfunded : Failure()
        data class Other(val message: String?) : Failure()
    }

    class Refused(val failure: Failure) : Exception()

    private fun other(message: String): Nothing = throw Refused(Failure.Other(message))

    /**
     * Every real contract call measured from the Safe's own address; the draft's
     * `callGasLimit` raised to the core's floor when that is higher. A call nobody
     * could measure leaves the relay's figure — and its existing guard — alone.
     */
    private suspend fun raisedToMeasuredFloor(
        draft: UserOpDraft,
        chainId: Int,
        account: String,
        calls: List<UserOpCall>,
    ): UserOpDraft {
        val toMeasure = userOpCallsToMeasure(calls)
        if (toMeasure.isEmpty()) return draft
        val measured = ArrayList<String>(toMeasure.size)
        for (index in toMeasure) {
            val call = calls[index.toInt()]
            val valueHex = if (call.value.startsWith("0x")) call.value else "0x" + java.math.BigInteger(call.value.ifBlank { "0" }).toString(16)
            val hex = measureCall(chainId, account, call.to, valueHex, call.data) ?: return draft
            measured += java.math.BigInteger(hex.removePrefix("0x"), 16).toString()
        }
        val raised = userOpRaiseCallGas(draft, measured, calls.size.toUInt())
        if (raised.callGasLimit != draft.callGasLimit) {
            VelaLog.event("userop.submit", "callGasLimit raised to the inner calls' own estimate", "relay" to draft.callGasLimit, "inner" to raised.callGasLimit)
        }
        return raised
    }

    /**
     * A page's message (spec 044): the Safe message hash under the Safe's
     * own domain is the passkey's challenge; the assertion is encoded as the
     * EIP-1271 envelope — `isValidSignature` verifies it on chain. One
     * ceremony, nothing submitted. Returns the signature hex.
     */
    suspend fun signMessage(chainId: Int, account: String, originalHash: ByteArray, signingStarted: () -> Unit = {}): String {
        val keys = accounts.keysOf(account)
        if (keys.isEmpty()) other("No passkey credential for the active account")
        val pinned = keys.first()
        val challenge = runCatching { safeMessageHash(originalHash, chainId.toULong(), account) }
            .getOrElse { other(it.message ?: "The message could not be hashed") }
        signingStarted()
        val (credentialId, transports, method) = routeFor(account, pinned)
        val assertion: Assertion = try {
            signer().sign(challenge, credentialId, transports, method)
        } catch (failure: PasskeyFailure) {
            if (failure.kind == FailureKind.Cancelled) throw Refused(Failure.PasskeyCancelled)
            other(failure.message ?: "the passkey ceremony failed")
        }
        val signature = runCatching {
            eip1271Signature(
                WebAuthnAssertion(
                    authenticatorData = SendExecutor.unhex(assertion.authenticatorDataHex),
                    clientDataJson = SendExecutor.unhex(assertion.clientDataJsonHex),
                    signatureDer = SendExecutor.unhex(assertion.signatureDerHex),
                ),
                assertion.credentialIdHex,
                keys,
            )
        }.getOrElse { other(it.message ?: "Failed to create signature") }
        return "0x" + signature.joinToString("") { b -> "%02x".format(b.toInt() and 0xff) }
    }

    /**
     * Assembles, signs once and submits; returns the accepted user-operation
     * hash (or the one already pending for this nonce). Throws [Refused].
     */
    suspend fun submit(
        chainId: Int,
        account: String,
        calls: List<UserOpCall>,
        gasFeeToken: String?,
        quotedFee: Quoted?,
        signingStarted: () -> Unit = {},
    ): String {
        val keys = accounts.keysOf(account)
        if (keys.isEmpty()) other("No passkey credential for the active account")
        val pinned = keys.first()
        val tempo = isChainWithoutNativeCoin(chainId.toUInt())
        val deployed = relay.isDeployed(chainId, account)
            ?: other("The network could not be reached. Please try again.")
        val nonce = if (deployed) {
            relay.nonce(chainId, account) ?: other("The account's nonce could not be read. Please try again.")
        } else {
            "0x0"
        }
        val floors = userOpFloors(chainId.toUInt(), deployed, (calls.size + 1).toUInt())
        val quoted = quotedFee
            ?.takeIf { quotedFeeUsable(it.amount, it.recipient) }
            ?: other("The fee quote has expired. Please review the updated fee and try again.")
        val feeToken = if (tempo) gasFeeToken ?: SendExecutor.TEMPO_DEFAULT_FEE_TOKEN else gasFeeToken
        val settled: UserOpFeeMode
        val placeholder: UserOpFeeMode
        if (tempo) {
            val collector = relay.accountInfo(chainId, account)?.feeRecipient()
                ?: other("The Tempo gas relayer is unavailable right now. Please try again.")
            if (!quoted.recipient.equals(collector, ignoreCase = true)) {
                other("The gas quote has expired. Please review the updated fee and try again.")
            }
            settled = UserOpFeeMode.Tempo(feeToken = feeToken!!, collector = collector, reimbursement = quoted.amount)
            placeholder = UserOpFeeMode.Tempo(feeToken = feeToken, collector = collector, reimbursement = "1")
        } else {
            settled = UserOpFeeMode.InBand(gasFeeToken = feeToken, amount = quoted.amount, recipient = quoted.recipient)
            placeholder = UserOpFeeMode.InBand(gasFeeToken = feeToken, amount = "1", recipient = account)
        }
        var draft = runCatching {
            userOpDraft(
                sender = account,
                nonce = nonce,
                deployed = deployed,
                keyHexes = keys.map { it.publicKeyHex },
                calls = calls,
                fee = placeholder,
                floors = floors,
            )
        }.getOrElse { other(it.message ?: "The operation could not be assembled.") }
        val hasContractCall = userOpHasContractCall(calls)
        when (val estimate = relay.estimateUserOpGas(chainId, userOpRelayJson(draft, feeToken.takeIf { tempo }))) {
            is RelayClient.EstimateAnswer.Estimated -> {
                draft = userOpApplyEstimate(
                    draft,
                    estimate.verificationGasLimit,
                    estimate.callGasLimit,
                    estimate.preVerificationGas,
                    floors,
                )
                draft = raisedToMeasuredFloor(draft, chainId, account, calls)
            }
            is RelayClient.EstimateAnswer.Refused, RelayClient.EstimateAnswer.Unreachable -> {
                VelaLog.event("userop.submit", "estimate failed, defaults", "contract" to hasContractCall)
                if (hasContractCall) {
                    other("Could not estimate gas for this transaction. The network may be busy — please try again.")
                }
            }
        }
        draft = userOpWithCalls(draft, calls, settled)
        val challenge = userOpSafeOpHash(draft, chainId.toUInt())
        signingStarted()
        val (credentialId, transports, method) = routeFor(account, pinned)
        val assertion: Assertion = try {
            signer().sign(challenge, credentialId, transports, method)
        } catch (failure: PasskeyFailure) {
            if (failure.kind == FailureKind.Cancelled) throw Refused(Failure.PasskeyCancelled)
            other(failure.message ?: "the passkey ceremony failed")
        }
        val signed = runCatching {
            userOpSign(
                draft,
                WebAuthnAssertion(
                    authenticatorData = SendExecutor.unhex(assertion.authenticatorDataHex),
                    clientDataJson = SendExecutor.unhex(assertion.clientDataJsonHex),
                    signatureDer = SendExecutor.unhex(assertion.signatureDerHex),
                ),
                assertion.credentialIdHex,
                keys,
            )
        }.getOrElse { other(it.message ?: "Failed to create signature") }
        return when (val answer = relay.sendUserOp(chainId, userOpRelayJson(signed, feeToken.takeIf { tempo }))) {
            is RelayClient.SubmitAnswer.Accepted -> answer.userOpHash
            is RelayClient.SubmitAnswer.Rejected -> {
                val message = relayErrorMessage(answer.errorJson)
                parseExistingUserOpHash(message)?.let { existing ->
                    VelaLog.event("userop.submit", "previous op pending", "hash" to existing.take(12))
                    return existing
                }
                throw Refused(
                    when (val rejection = classifyRelayRejection(message)) {
                        RelayRejection.RelayerUnavailable -> Failure.RelayerUnavailable
                        RelayRejection.BundlerUnderfunded -> Failure.BundlerUnderfunded
                        is RelayRejection.Other -> Failure.Other(rejection.message.ifBlank { null })
                    },
                )
            }
            RelayClient.SubmitAnswer.Unreachable -> other("The gas relayer could not be reached. Please try again.")
        }
    }
}

File: app-ios/VelaWallet/VelaWallet/Core/UserOpSpine.swift
//
//  UserOpSpine.swift
//  VelaWallet
//
//  The submit spine — ONE implementation for a person's own transfer and a
//  dApp's transaction.
//
//  Ported from `app-android/.../feature/send/core/UserOpSpine.kt` (spec 044
//  T028), which is the desktop's `executor::user_op::submit`. The order below
//  is the file's content and is not to be improvised:
//
//      keys → the Tempo branch → isDeployed → nonce → floors → the quote gate
//      → a placeholder fee leg → draft → estimate → apply → the SETTLED fee
//      leg → the SafeOp hash IS the challenge → the assertion → the envelope
//      → the relay
//
//  The assembly is the core's (`user_op_*` through uniffi); the transports are
//  `RelayClient`'s; the one seam the core cannot have is `UserOpSigner`.
//  Nothing here prices, validates or classifies beyond what the core exports.
//
//  ## Two refusals happen BEFORE any prompt
//
//  A deployed wallet whose nonce could not be read, and a batch carrying a real
//  contract call whose gas could not be estimated. A passkey prompt on an
//  operation the relay must reject is a prompt wasted — and on this platform it
//  is also a Face ID sheet somebody has to dismiss for nothing.
//
//  ## Why `signMessage` lives here and 052 never calls it
//
//  A dApp's `personal_sign` (spec 053) signs the Safe message hash and wraps
//  the assertion in an EIP-1271 envelope. It belongs beside the transfer for
//  the same reason the two share a file on every other client: the sheet that
//  says what will happen and the code that makes it happen must not become two
//  opinions. Splitting it across two cuts is exactly how they would.
//

import Foundation
import VelaCore

extension UserOpSpine.AccountPort {
    /// Nothing known: the ceremony routes as it always did.
    func keyRoutesJson(of address: String) async -> String { "[]" }
}

@MainActor
final class UserOpSpine {

    /// The displayed fee, signed verbatim.
    struct Quoted {
        let amount: String
        let recipient: String
    }

    /// The account store, as the submit path reads it.
    protocol AccountPort {
        /// The wallet's founding keys for `address`, pinned key first; empty
        /// when unknown. **All** of them: the Safe's address is a function of
        /// the whole set, so a signature packed against a subset verifies
        /// against a different account.
        func keys(of address: String) async -> [WalletKeyRecord]
        /// The pinned key's stored transports and method, for the ceremony's
        /// routing.
        func routing(of address: String) async -> (transports: String, method: KeyMethod)
        /// Every founding key's credential id and stored transports, as JSON for
        /// the core's `signRoute` — `[{credential_id, transports}]`.
        func keyRoutesJson(of address: String) async -> String
    }

    enum Failure: Equatable {
        case passkeyCancelled
        case relayerUnavailable
        case bundlerUnderfunded
        case other(String?)
    }

    struct Refused: Error {
        let failure: Failure
    }

    private let relay: RelayClient
    private let accounts: AccountPort
    private let signer: () -> UserOpSigner
    /// `eth_estimateGas({from, to, value, data})` on one chain — the gas hex, or
    /// `nil` when nobody answered. The inner calls' own measurement (spec 062):
    /// the relay estimates `callGasLimit` against the SENDER, and an undeployed
    /// Safe has no code there, so its figure is 21k plus calldata whatever the
    /// call does. Measured from the Safe's address instead, a codeless account
    /// estimates like any other. The rule is the core's (`userOpRaiseCallGas`).
    private let measureCall: (_ chainId: Int, _ from: String, _ to: String, _ valueHex: String, _ data: String) async -> String?

    init(
        relay: RelayClient,
        accounts: AccountPort,
        signer: @escaping () -> UserOpSigner,
        measureCall: @escaping (Int, String, String, String, String) async -> String? = { _, _, _, _, _ in nil }
    ) {
        self.relay = relay
        self.accounts = accounts
        self.signer = signer
        self.measureCall = measureCall
    }

    /// Every real contract call measured from the Safe's own address; the draft's
    /// `callGasLimit` raised to the core's floor when that is higher. A call nobody
    /// could measure leaves the relay's figure — and its existing guard — alone.
    private func raisedToMeasuredFloor(
        _ draft: UserOpDraft, chainId: Int, account: String, calls: [UserOpCall]
    ) async -> UserOpDraft {
        guard let toMeasure = try? userOpCallsToMeasure(calls: calls), !toMeasure.isEmpty else { return draft }
        var measured: [String] = []
        for index in toMeasure {
            let call = calls[Int(index)]
            let valueHex = call.value.hasPrefix("0x") ? call.value : "0x" + (Self.decimalToHex(call.value) ?? "0")
            guard let hex = await measureCall(chainId, account, call.to, valueHex, call.data),
                  let gas = UInt64(hex.dropFirst(hex.hasPrefix("0x") ? 2 : 0), radix: 16)
            else { return draft }
            measured.append(String(gas))
        }
        return (try? userOpRaiseCallGas(draft: draft, measured: measured, callCount: UInt32(calls.count))) ?? draft
    }

    /// A decimal wei string as bare hex. Values here are small enough for
    /// `UInt64`… except when they are not — then the measurement is skipped.
    private static func decimalToHex(_ decimal: String) -> String? {
        decimal.isEmpty ? "0" : UInt64(decimal).map { String($0, radix: 16) }
    }

    private func other(_ message: String) -> Refused { Refused(failure: .other(message)) }

    // MARK: - A page's message (spec 053; ported now, called then)

    /// The Safe message hash under the Safe's own domain is the passkey's
    /// challenge; the assertion becomes the EIP-1271 envelope, which
    /// `isValidSignature` verifies on chain. One ceremony, nothing submitted.
    func signMessage(
        chainId: Int,
        account: String,
        originalHash: Data,
        signingStarted: () -> Void = {}
    ) async throws -> String {
        let keys = await accounts.keys(of: account)
        guard let pinned = keys.first else {
            throw other("No passkey credential for the active account")
        }
        let challenge: Data
        do {
            challenge = try safeMessageHash(
                originalHash: originalHash, chainId: UInt64(chainId), safeAddress: account
            )
        } catch {
            throw other("The message could not be hashed")
        }
        signingStarted()
        let assertion = try await assert(account: account, pinned: pinned, challenge: challenge)
        do {
            let signature = try eip1271Signature(
                assertion: Self.webAuthn(assertion),
                credentialId: assertion.credentialIdHex,
                keys: keys
            )
            return "0x" + signature.map { String(format: "%02x", $0) }.joined()
        } catch {
            throw other("Failed to create signature")
        }
    }

    // MARK: - The submit

    /// Assembles, signs once and submits; answers the accepted user-operation
    /// hash — or the one already pending for this nonce, which is an idempotent
    /// re-submit rather than a second spend.
    func submit(
        chainId: Int,
        account: String,
        calls: [UserOpCall],
        gasFeeToken: String?,
        quotedFee: Quoted?,
        signingStarted: () -> Void = {}
    ) async throws -> String {
        let keys = await accounts.keys(of: account)
        guard let pinned = keys.first else {
            throw other("No passkey credential for the active account")
        }
        let tempo = isChainWithoutNativeCoin(chainId: UInt32(chainId))

        guard let deployed = await relay.isDeployed(chainId: chainId, address: account) else {
            throw other("The network could not be reached. Please try again.")
        }
        // Read before the prompt, and fatal: an operation with the wrong nonce
        // is one the relay must reject, and a Face ID sheet raised for it is a
        // sheet somebody dismisses for nothing.
        let nonce: String
        if deployed {
            guard let read = await relay.nonce(chainId: chainId, sender: account) else {
                throw other("The account's nonce could not be read. Please try again.")
            }
            nonce = read
        } else {
            nonce = "0x0"
        }

        let floors = userOpFloors(
            chainId: UInt32(chainId), deployed: deployed, subCalls: UInt32(calls.count + 1)
        )

        // The displayed-equals-signed gate. `quoted_fee` is present ⇔ in-band,
        // and the core's own predicate decides whether it is still usable.
        guard let quoted = quotedFee, quotedFeeUsable(amount: quoted.amount, recipient: quoted.recipient) else {
            throw other("The fee quote has expired. Please review the updated fee and try again.")
        }

        let feeToken = tempo ? (gasFeeToken ?? Self.tempoDefaultFeeToken) : gasFeeToken
        let settled: UserOpFeeMode
        let placeholder: UserOpFeeMode
        if tempo {
            guard let collector = await relay.accountInfo(chainId: chainId, safe: account)?.feeRecipient else {
                throw other("The Tempo gas relayer is unavailable right now. Please try again.")
            }
            guard quoted.recipient.caseInsensitiveCompare(collector) == .orderedSame else {
                throw other("The gas quote has expired. Please review the updated fee and try again.")
            }
            settled = .tempo(feeToken: feeToken!, collector: collector, reimbursement: quoted.amount)
            placeholder = .tempo(feeToken: feeToken!, collector: collector, reimbursement: "1")
        } else {
            settled = .inBand(gasFeeToken: feeToken, amount: quoted.amount, recipient: quoted.recipient)
            // The placeholder exists so the estimate runs against an operation
            // the same SHAPE as the real one — one more leg, non-zero amounts —
            // without committing to the quoted figure before it is checked.
            placeholder = .inBand(gasFeeToken: feeToken, amount: "1", recipient: account)
        }

        var draft: UserOpDraft
        do {
            draft = try userOpDraft(
                sender: account,
                nonce: nonce,
                deployed: deployed,
                keyHexes: keys.map(\.publicKeyHex),
                calls: calls,
                fee: placeholder,
                floors: floors
            )
        } catch {
            throw other("The operation could not be assembled.")
        }

        let hasContractCall = (try? userOpHasContractCall(calls: calls)) ?? false
        let relayJson: String
        do {
            relayJson = try userOpRelayJson(draft: draft, feeToken: tempo ? feeToken : nil)
        } catch {
            throw other("The operation could not be encoded.")
        }
        switch await relay.estimateUserOpGas(chainId: chainId, opJson: relayJson) {
        case .estimated(let verification, let call, let preVerification):
            if let applied = try? userOpApplyEstimate(
                draft: draft,
                verificationGasLimit: verification,
                callGasLimit: call,
                preVerificationGas: preVerification,
                floors: floors
            ) {
                draft = await raisedToMeasuredFloor(applied, chainId: chainId, account: account, calls: calls)
            }
        case .refused, .unreachable:
            // A plain transfer can ride the floors. A batch carrying a real
            // contract call cannot: its gas is unknowable without the estimate,
            // and submitting anyway spends a prompt on a rejection.
            if hasContractCall {
                throw other("Could not estimate gas for this transaction. The network may be busy — please try again.")
            }
        }

        do {
            draft = try userOpWithCalls(draft: draft, calls: calls, fee: settled)
        } catch {
            throw other("The operation could not be assembled.")
        }

        let challenge: Data
        do {
            challenge = try userOpSafeOpHash(draft: draft, chainId: UInt32(chainId))
        } catch {
            throw other("The operation could not be hashed.")
        }
        signingStarted()
        let assertion = try await assert(account: account, pinned: pinned, challenge: challenge)

        let signed: UserOpDraft
        do {
            signed = try userOpSign(
                draft: draft,
                assertion: Self.webAuthn(assertion),
                credentialId: assertion.credentialIdHex,
                keys: keys
            )
        } catch {
            throw other("Failed to create signature")
        }

        let signedJson: String
        do {
            signedJson = try userOpRelayJson(draft: signed, feeToken: tempo ? feeToken : nil)
        } catch {
            throw other("The operation could not be encoded.")
        }

        switch await relay.sendUserOp(chainId: chainId, opJson: signedJson) {
        case .accepted(let hash):
            return hash
        case .unreachable:
            throw other("The gas relayer could not be reached. Please try again.")
        case .rejected(let errorJson):
            let message = relayErrorMessage(errorJson: errorJson)
            // Already pending for this nonce: the same operation, not a second
            // one. Answering its hash lets the tracker follow what is really on
            // the wire instead of submitting a duplicate.
            //
            // **The RAW json is searched first, and that is a deviation from
            // the other clients.** `relay_error_message` is a translator, not a
            // pass-through: a message matching a known rung ("AA25", "invalid
            // account nonce") is REPLACED wholesale by its human sentence, and
            // the relay's `[existingHash:0x…]` marker goes with it. A second
            // submit for the same nonce is exactly an AA25 condition, so the
            // one case where the marker matters most is the one where the
            // translation can eat it. Android and the desktop parse only the
            // translated text; this is reported rather than quietly diverged.
            if let existing = parseExistingUserOpHash(message: errorJson)
                ?? parseExistingUserOpHash(message: message) {
                return existing
            }
            switch classifyRelayRejection(message: message) {
            case .relayerUnavailable: throw Refused(failure: .relayerUnavailable)
            case .bundlerUnderfunded: throw Refused(failure: .bundlerUnderfunded)
            case .other(let text): throw Refused(failure: .other(text.isEmpty ? nil : text))
            }
        }
    }

    // MARK: - The one ceremony

    /// The person's "Sign with" choice for the request in hand — `auto` unless a
    /// signing sheet says otherwise (the spine is shared with Send, which never
    /// sets it). WHICH key that pins, and how it is reached, is the core's.
    var signMethod: () -> String = { "auto" }

    /// The credential a ceremony is pinned to, its transports and method: the
    /// person's choice when they made one, the stored route otherwise.
    private func route(account: String, first: WalletKeyRecord) async -> (credentialId: String, transports: String, method: KeyMethod) {
        let chosen = signMethod()
        if chosen != "auto",
           let json = signRoute(deviceKeysJson: await accounts.keyRoutesJson(of: account), method: chosen),
           let picked = (try? JSONSerialization.jsonObject(with: Data(json.utf8))) as? [String: Any],
           let credential = picked["credential_id"] as? String, !credential.isEmpty,
           let method = (picked["method"] as? String).flatMap(KeyMethod.init(rawValue:))
        {
            return (credential, picked["transports"] as? String ?? "", method)
        }
        let stored = await accounts.routing(of: account)
        return (first.credentialId, stored.transports, stored.method)
    }

    private func assert(
        account: String,
        pinned: WalletKeyRecord,
        challenge: Data
    ) async throws -> Assertion {
        let routing = await route(account: account, first: pinned)
        do {
            return try await signer().sign(
                challenge: challenge,
                credentialIdHex: routing.credentialId,
                transports: routing.transports,
                method: routing.method
            )
        } catch let failure as PasskeyFailure {
            if failure.kind == .cancelled { throw Refused(failure: .passkeyCancelled) }
            throw other(failure.message.isEmpty ? "the passkey ceremony failed" : failure.message)
        } catch is CancellationError {
            throw Refused(failure: .passkeyCancelled)
        }
    }

    private static func webAuthn(_ assertion: Assertion) -> WebAuthnAssertion {
        WebAuthnAssertion(
            authenticatorData: unhex(assertion.authenticatorDataHex),
            clientDataJson: unhex(assertion.clientDataJsonHex),
            signatureDer: unhex(assertion.signatureDerHex)
        )
    }

    /// `fee_policy::TEMPO_DEFAULT_FEE_TOKEN` — pathUSD.
    static let tempoDefaultFeeToken = "0x20c0000000000000000000000000000000000000"

    static func unhex(_ text: String) -> Data {
        var body = Substring(text)
        if body.hasPrefix("0x") || body.hasPrefix("0X") { body = body.dropFirst(2) }
        var out = Data()
        var index = body.startIndex
        while index < body.endIndex {
            let next = body.index(index, offsetBy: 2, limitedBy: body.endIndex) ?? body.endIndex
            guard next > index, let byte = UInt8(body[index..<next], radix: 16) else { break }
            out.append(byte)
            index = next
        }
        return out
    }
}
